### PR TITLE
Add JSONAsStringRowInputFormat

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add new input format `JSONAsString` that accepts a sequence of JSON objects separated by newlines, spaces and/or commas.

Detailed description / Documentation draft:

### JSONAsString input format
This format parses a sequence of JSON objects separated by newlines, spaces and/or comma.
Each JSON object is parsed as a whole to string. This format can only parse a table with single field of type String.

### CODE REVIEW

File: tests/queries/0_stateless/01232_json_as_string_format.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS json_as_string;
+CREATE TABLE json_as_string (field String) ENGINE = Memory;
+
+INSERT INTO json_as_string FORMAT JSONAsString {"id" : 1, "date" : "01.01.2020", "string" : "123{{{\"\\", "array" : [1, 2, 3], "map": {"a" : 1, "b" : 2, "c" : 3}} , {"id" : 2, "date" : "01.02.2020", "string" : "{another\" string}}", "array" : [3, 2, 1], "map" : {"z" : 1, "y" : 2, "x" : 3}} {"id" : 3, "date" : "01.03.2020", "string" : "one more string", "array" : [3,1,2], "map" : {"{" : 1, "}}" : 2}}
```
Comment (filimonov) [inline tests/queries/0_stateless/01232_json_as_string_format.sql:4]: Test with multiline json?

**Comment by alexey-milovidov** [comment]
Minor:
```
src/Processors/Formats/Impl/JSONAsStringRowInputFormat.cpp:127:16: error: redundant boolean literal in conditional return statement [readability-simplify-boolean-expr,-warnings-as-errors]
2020-04-30 21:54:36         return false;
2020-04-30 21:54:36 ~~~~~~~~~~~~~~~^~~~~~
```

File: src/IO/PeekableReadBuffer.cpp [hunk 2]
```diff
@@ -245,6 +245,7 @@ void PeekableReadBuffer::makeContinuousMemoryFromCheckpointToPos()
     memcpy(memory.data() + peeked_size, sub_buf.position(), bytes_to_append);
     sub_buf.position() = pos;
     peeked_size += bytes_to_append;
+    BufferBase::set(memory.data(), peeked_size, peeked_size);
```
Comment (Avogar) [inline src/IO/PeekableReadBuffer.cpp:248]: After buf.makeContinuousMemoryFromCheckpointToPos() checkpoint and buf.position() were in different buffers. Now they are in the same. (This fix was made in agreement with @tavplubix)

**Review by alexey-milovidov — APPROVED** [review]

File: src/Processors/Formats/Impl/JSONAsStringRowInputFormat.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
+#include <Processors/Formats/Impl/JSONAsStringRowInputFormat.h>
+#include <Formats/JSONEachRowUtils.h>
+#include <common/find_symbols.h>
+#include <IO/ReadHelpers.h>
+
+namespace DB
+{
+
+extern bool fileSegmentationEngineJSONEachRowImpl(ReadBuffer & in, DB::Memory<> & memory, size_t min_chunk_size);
```
Comment (azat) [inline src/Processors/Formats/Impl/JSONAsStringRowInputFormat.cpp:9]: is redundant, since the declaration is in `JSONEachRowUtils.h`

Comment (alexey-milovidov) [reply src/Processors/Formats/Impl/JSONAsStringRowInputFormat.cpp:9]: Ok, thank you!

File: tests/queries/0_stateless/01232_json_as_string_format.sql (hunk 1 above)
Comment (filimonov) [reply tests/queries/0_stateless/01232_json_as_string_format.sql]: Thanks!

**Comment by qoega** [comment]
For documentation: https://github.com/ClickHouse/ClickHouse/pull/10607/files#diff-fef10815cbf8ba0462108a68efe8f4c4R12-R14

**Comment by vladimir-golovchenko** [comment]
@ramazanpolat I get the same exception for version 20.6.4.44. It looks like need to add the bug.

**Comment by Avogar** [comment]
> ```
> CREATE TABLE doc(s String) Engine=Memory;
> INSERT INTO doc FORMAT JSONAsString '{"a":1}';
> ```
> 
> Gives this:
> 
> ```
> Code: 117, e.displayText() = DB::Exception: JSON object must begin with '{'.: (at row 1)
> , Stack trace (when copying this message, always include the lines below):
> 
> 0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x12400be0 in /usr/bin/clickhouse
> 1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xa2433fd in /usr/bin/clickhouse
> 2. ? @ 0xfc7af97 in /usr/bin/clickhouse
> 3. DB::JSONAsStringRowInputFormat::readRow(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, DB::RowReadExtension&) @ 0xfc7a8e7 in /usr/bin/clickhouse
> 4. DB::IRowInputFormat::generate() @ 0xfc2d7a1 in /usr/bin/clickhouse
> 5. DB::ISource::work() @ 0xfba50db in /usr/bin/clickhouse
> 6. DB::InputStreamFromInputFormat::readImpl() @ 0xfb69edd in /usr/bin/clickhouse
> 7. DB::IBlockInputStream::read() @ 0xede92cd in /usr/bin/clickhouse
> 8. DB::ParallelParsingBlockInputStream::parserThreadFunction(std::__1::shared_ptr<DB::ThreadGroupStatus>, unsigned long) @ 0xfb6e7f8 in /usr/bin/clickhouse
> 9. ? @ 0xfb6f554 in /usr/bin/clickhouse
> 10. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xa272a57 in /usr/bin/clickhouse
> 11. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const @ 0xa2731ca in /usr/bin/clickhouse
> 12. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa271f67 in /usr/bin/clickhouse
> 13. ? @ 0xa2704a3 in /usr/bin/clickhouse
> 14. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
> 15. __clone @ 0x121a3f in /lib/x86_64-linux-gnu/libc-2.27.so
>  (version 20.6.4.44 (official build))
> 
> Exception on client:
> Code: 117. DB::Exception: JSON object must begin with '{'.: (at row 1)
> ```
> 
> What am I doing wrong?

I guess you need to remove ' symbol from the edges of the JSON object. So, try this:
```
CREATE TABLE doc(s String) Engine=Memory;
INSERT INTO doc FORMAT JSONAsString {"a":1};
```

**Comment by RonSour** [comment]
Kafka and JSONAsString any example?

**Pull request merged**